Implement EmmyLua alias builder

C3pa opened the pull request:
This builder will create EmmyLua `---@alias` annotations for tes3.* enumeration tables. Here I appended how the built file looks like: [aliases.zip](https://github.com/MWSE/MWSE/files/8456759/aliases.zip). This will allow us to specify which enumeration a certain field accepts. For example, in tes3baseObject.objectType we can now change this to "tes3.baseObject". Moreover, this is even better for function definitions, where we can do can state which `tes3.` enumeration the field maps to. This will provide really nice autocomplete. Here are some examples:
 

![---@alias 1](https://user-images.githubusercontent.com/41503714/162568910-61a4d6ed-d00b-429c-ba55-1e39e3be5015.png)
![---@alias 2](https://user-images.githubusercontent.com/41503714/162568912-6f1eda1e-7d8c-432b-bd25-0b329abf9b52.png)
![---@alias 3](https://user-images.githubusercontent.com/41503714/162568915-06191530-6196-4052-b870-a479c1556b4d.png)


I had an idea previously, to achieve mostly the same purpose by declaring `---@class` at the top of every `tes3.*` enumeration. It doesn't give that good autocomplete in the end. Anyway, if anyone else wants to try out the `---@class` approach, here I provide the files I used to compare both: [Test Files.zip](https://github.com/MWSE/MWSE/files/8456762/Test.Files.zip). A note about using these: you will need to remove all the files copied from one folder to try the other approach. 

So, a list of things that need to be resolved before merging this is:
- [ ] Is the blacklist approach acceptable? ~~To reduce the amount of entries in the blacklist, I could move/implement with table.invert() the reversed tables that map IDs to names (attributeName.lua and similar) init.lua.~~
- [ ] Are we happy with the name of the builder?
- [ ] Are we happy with the output file destination?
- [ ] Discuss if I should integrate this builder better with the other builders.
- [ ] Discuss the position of this builder in the build queue. Is it okay for it to be the first?	
- [ ] Update relevant parts of the documentation to use the new feature.

C3pa commented:
I don't think that an automatic builder for `tes3.` enumeration table -> mkdocs is needed, because of the changes Sephumbra is about to commit. They said that they are going to add descriptions to the entries, but we don't have such descriptions in `tes3.*` files themselves.

pierresteeg commented:
> I don't think that an automatic builder for `tes3.` enumeration table -> mkdocs is needed, because of the changes Sephumbra is about to commit. They said that they are going to add descriptions to the entries, but we don't have such descriptions in `tes3.*` files themselves.

Automating documentation for these tables will not be straight forward. We have multiple exceptions with either irregular naming, nested tables, or additional columns that may be of interest like descriptions/default values in case of GMSTs and so on. We would also need to mark/define which tables to generate docs for since its just inside the tes3 table.

Would be easier if we were able to insert more info into these tables, but thats not really an option at this point.

pierresteeg commented:
I would suggest not blacklisting any tables like scanCodeToNumber/specializationNames and so on. Those tables are functionally identical to the ones aliases are being created for and generating alias annotations wont hurt anybody, but rather reduce confusion in the future when a function wants to make use of those and autocompletion is not normally generated.

Blacklisting for something like init.lua makes sense to me, although we have to keep this in mind in case we want to integrate something else into the tes3 folder.

I dont really know enough about the rest of the doc builder system to chime in on the other points.

C3pa commented:
This doesn't work anymore. A new version 3.0.0 of Lua extension landed today. It worked in 2.6.8. In 2.6.8. when you start typing page parameter, after `=` sign you get the autocomplete suggestions as shown on the screenshot above. In 3.0.0. you get those suggestions inside quotation marks. Obviously, `tes3.dialoguePage.voice.hello` isn't the same as `"tes3.dialoguePage.voice.hello"`.

I'll see how to fix this.

C3pa commented:
Sephumbra, I changed the blacklist as you requested. Also, I made it work with the latest version of Sumneko's Lua plugin. To test this in action, drop the built aliases file in your `MWSE\core\meta\class` folder, and I modified tes3.lua meta file so it uses these annotations. [Data Files.zip](https://github.com/MWSE/MWSE/files/9023378/Data.Files.zip) Just start typing objectType field on tes3.createObject or page and type in tes3.findDialogue. Currently, only those 2 functions are modified for testing purposes. I will change all the appropriate parts of the documentation to use this feature before merging.

pierresteeg commented on autocomplete/.vscode/tasks.json:
```diff
@@ -45,6 +55,7 @@
 				"isDefault": true
 			},
 			"dependsOn": [
+				"Build EmmyLua Aliases",
```
Since all of the autocomplete is technically EmmyLua, I don't think this needs specifying. Perhaps this could also just be put into the regular builder instead of a separate step. I have never looked at the builder though and might not be the best to give my input on this.

pierresteeg commented on misc/package/Data Files/MWSE/core/meta/class/aliases.lua:
```diff
@@ -0,0 +1,3305 @@
+-- This file is autogenerated. Do not edit this file manually. Your changes will be ignored.
```
We generally split most things into their own files. Is there a reason we went for a single file this time instead of splitting them up into actionFlag.lua and similar this time?

pierresteeg reviewed (commented):
Not the best person to give input on this since I have no idea about the builder, but here's the things that jumped out at me.

pierresteeg commented:
I will give the result of this a try *probably* this weekend. The built aliases looked fine to me from the output file though.

C3pa commented:
As Sumneko is feature-loading his plugin we could have an easier time achieving this thing. Here is the [related issue](https://github.com/sumneko/lua-language-server/issues/1255) upstream. I'll stop development here until we see how the new feature plays out.

C3pa commented:
This can be used as a starting point for some future work.
